cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MainOutputCommitter.java

```diff
@@ -0,0 +1,286 @@
+    MapReduceTaskContextProvider taskContextProvider = classLoader.getTaskContextProvider();
+    Injector injector = taskContextProvider.getInjector();
+
+    cConf = injector.getInstance(CConfiguration.class);
+    MapReduceContextConfig contextConfig = new MapReduceContextConfig(jobContext.getConfiguration());
+
+    ProgramId programId = contextConfig.getProgramId();
+    LOG.info("Setting up for MapReduce job: namespaceId={}, applicationId={}, program={}, runid={}",
+             programId.getNamespace(), programId.getApplication(), programId.getProgram(),
+             ProgramRunners.getRunId(contextConfig.getProgramOptions()));
+
+    RetryStrategy retryStrategy =
+      SystemArguments.getRetryStrategy(contextConfig.getProgramOptions().getUserArguments().asMap(),
+                                       contextConfig.getProgramId().getType(), cConf);
+    this.txClient = new RetryingLongTransactionSystemClient(injector.getInstance(TransactionSystemClient.class),
+                                                            retryStrategy);
+
+    // We start long-running tx to be used by mapreduce job tasks.
+    this.transaction = txClient.startLong();
+
+    // Write the tx somewhere, so that we can re-use it in mapreduce tasks
+    Path txFile = getTxFile(configuration, jobContext.getJobID());
+    FileSystem fs = txFile.getFileSystem(configuration);
+    try (FSDataOutputStream fsDataOutputStream = fs.create(txFile, false)) {
+      fsDataOutputStream.write(new TransactionCodec().encode(transaction));
+    }
+
+    // we can instantiate the TaskContext after we set the tx above. It's used by the operations below
+    taskContext = taskContextProvider.get(taskAttemptContext);
+    this.outputs = Outputs.transform(contextConfig.getOutputs(), taskContext);
+
+    super.setupJob(jobContext);
+  }
+
+  // returns a Path to a file in the staging directory of the MapReduce
+  static Path getTxFile(Configuration configuration, @Nullable JobID jobID) throws IOException {
+    Path stagingDir = MRApps.getStagingAreaDir(configuration, UserGroupInformation.getCurrentUser().getShortUserName());
+    int appAttemptId = configuration.getInt(MRJobConfig.APPLICATION_ATTEMPT_ID, 0);
+    // following the pattern of other files in the staging dir
+    String jobId = jobID != null ? jobID.toString() : configuration.get(MRJobConfig.ID);
+    return new Path(stagingDir, jobId + "/"  + jobId + "_" + appAttemptId + "_tx-file");
+  }
+
+  @Override
+  public void commitJob(JobContext jobContext) throws IOException {
+    super.commitJob(jobContext);
+
+    onFinish(jobContext, true);
+    commitTx();
+  }
+
+  private void commitTx() throws IOException {
+    try {
+      LOG.debug("Committing MapReduce Job transaction: {}", transaction.getWritePointer());
+
+      // "Commit" the data event topic by publishing an empty message.
+      // Need to do it with the raw MessagingService.
```

Comment:
I had moved it from MapReduceRuntimeService, where it was happening before the tx is committed: https://github.com/caskdata/cdap/pull/9442/files/87b2aac25919f27c853478505a991abd9bcefbd0#diff-c905d8c90de3014351ec78f30d55b0b5L433

@chtyim any suggestion if this needs to be moved to after the tx commit?